src/kernel-install/ukify.py

```diff
@@ -0,0 +1,233 @@
+#!/usr/bin/python
+# SPDX-License-Identifier: LGPL-2.1+
+
+import argparse
+import dataclasses
+import itertools
+import pathlib
+import os
+import shlex
+import subprocess
+import tempfile
+import typing
+
+efi_arches = {
+        # host_arch: efi_arch
+        'x86'    : 'ia32',
+        'x86_64' : 'x64',
+        'arm'    : 'arm',
+        'aarch64': 'aa64',
+        'riscv64': 'riscv64',
+}
+
+def guess_efi_arch():
+    arch = os.uname().machine
+    efi_arch = efi_arches[arch]
+
+    if arch == 'x86_64':
+        try:
+            size = open('/sys/firmware/efi/fw_platform_size').read().strip()
+        except FileNotFoundError:
+            pass
+        else:
+            if int(size) == 32:
+                efi_arch = 'ia32'
+
+    print(f'Host arch {arch!r}, efi arch {efi_arch!r}')
+    return efi_arch
+
+def shell_join(cmd):
+    return ' '.join(shlex.quote(str(x)) for x in cmd)
+
+@dataclasses.dataclass
+class Section:
+    name: str
+    content: str
+    offset: int
+    tmpfile: typing.Optional[typing.IO] = None
+
+    @classmethod
+    def create(cls, name, content, offset):
+        if isinstance(content, str):
+            tmp = tempfile.NamedTemporaryFile(mode='wt', prefix=f'tmp{name}')
+            tmp.write(content)
+            tmp.flush()
+            content = tmp.name
+        else:
+            tmp = None
+
+        return cls(name, content, offset, tmpfile=tmp)
+
+    @classmethod
+    def parse_arg(cls, s):
+        parts = s.split(':')
+        if len(parts) < 2 or len(parts) > 3:
+            raise ValueError(f'Cannot parse section spec: {s!r}')
+        name, content = parts[:2]
```

Comment:
This should warn if the section name is longer than 8 chars. It'd be legal PE section name, but our stub cannot locate those. If this feature is used when prototyping new sections this can be caught early.

Reply:
I added a check now. https://learn.microsoft.com/en-us/windows/win32/debug/pe-format#section-table-section-headers says that anything utf-8 goes, but I think it's reasonable to limit ourselves to ascii now.